<a href="https://colab.research.google.com/github/eren-darici/STAT6620/blob/main/6620_HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

$$
\begin{array}{c}
\text{Homework 3} \\
\text{Eren Darici} \\
\text{STAT 6620} \\
\text{Date: 9/17/2025}
\end{array}
$$


In [3]:
import torch
import pandas as pd

In [4]:
# Read the lungfunction.txt file, delimited by space or tab, with specified column names
df = pd.read_csv('lungfunction.txt', sep='\s+', names=["age", "FEV", "ht", "sex", "smoke"])

# Display the first few rows of the DataFrame
df.count()

<>:2: SyntaxWarning: invalid escape sequence '\s'
<>:2: SyntaxWarning: invalid escape sequence '\s'
/tmp/ipython-input-1764796367.py:2: SyntaxWarning: invalid escape sequence '\s'
  df = pd.read_csv('lungfunction.txt', sep='\s+', names=["age", "FEV", "ht", "sex", "smoke"])


age      654
FEV      654
ht       654
sex      654
smoke    654
dtype: int64

In [11]:
# X = age, Y = FEV
X = torch.tensor(df.age.values, dtype=torch.float32).view(-1, 1)
X = torch.cat((torch.ones_like(X), X), dim=1)
Y = torch.tensor(df.FEV.values, dtype=torch.float32).view(-1, 1)

# Part a.
$$
  X^{T}X = X^T \, X
$$

In [14]:
# Part a.
XtX = X.T @ X

# Display the result
print("X'X:")
display(XtX)

X'X:


tensor([[  654.,  6495.],
        [ 6495., 70201.]])

$$
X^TX = \begin{bmatrix}
  654. &&  6495.\\
  6495. && 70201.
\end{bmatrix}
$$

# Part b.
$$
X^TXb=X^T \, Y \\
b=(X^TX)^{-1}X^TY
$$

In [46]:
# Part b.
# XtXb = XtY ->  b = (XtX)^-1 @ XtY
XtY = X.T @ Y

# Display the result
print("X'Y:")
display(XtY)

# b = (XtX)^-1 @ XtY
b = torch.linalg.inv(XtX) @ XtY

# Display the result
print("b:")
display(b)

X'Y:


tensor([[ 1724.4535],
        [18391.0488]])

b:


tensor([[0.4316],
        [0.2220]])

$$
X^TY =
\begin{bmatrix}
1724.4535\\
18391.0488
\end{bmatrix}
\\
\\
b =\begin{bmatrix}
b_0\\
b_1
\end{bmatrix} = \begin{bmatrix}
0.4316\\
0.2220
\end{bmatrix}
$$

# Part c.
$$
\hat{Y} = Xb \\
e = Y - \hat{Y} \\
SSE = e^Te \\
MSE = \frac{e^Te}{n-p} \\
$$
where $n = 654$ observations and $p = 2$

In [51]:
# Part c.
Yhat = X @ b
e = Y - Yhat

SSE = (e.T @ e).item()
print("SSE:")
display(SSE)

n = len(Y)
p = X.shape[1]

MSE = SSE / (n-p)
print("MSE:")
display(MSE)

SSE:


210.00064086914062

MSE:


0.32208687249868195

$$
SSE = 210.00064086914062 \\
MSE = 0.32208687249868195
$$

# Part (d)
$$
\text{Cov}(\hat{B}) = \text{MSE} \cdot (X^T X)^{-1} \\
s(b_1) = \sqrt{\text{Cov}(\hat{B})_{1,1}}
$$


In [53]:
# Part d.
cov_B = MSE * torch.linalg.inv(XtX)
print("Cov(B):")
display(cov_B)
sb1 = torch.sqrt(cov_B[1,1])
print("sb1:")
display(sb1)

Cov(B):


tensor([[ 6.0677e-03, -5.6138e-04],
        [-5.6138e-04,  5.6527e-05]])

sb1:


tensor(0.0075)

$$
\text{Cov}(\hat{B}) = \begin{bmatrix}
6.0677e-03 &-5.6138e-04\\
        -5.6138e-04 & 5.6527e-05\\
\end{bmatrix} \\
s(b_1) = \sqrt{\text{Cov}(\hat{B})_{1,1}} = \sqrt{(6.0677e-03} \approx 0.0075
$$
